resilience4j-ratelimiter/src/main/java/io/github/resilience4j/ratelimiter/internal/SemaphoreBasedRateLimiter.java

```diff
@@ -144,13 +145,35 @@ public boolean acquirePermission() {
     }
 
     /**
-     * {@inheritDoc}
-     * SemaphoreBasedRateLimiter is totally blocking by it's nature. So this non-blocking API isn't supported.
-     * It will return negative numbers all the time.
+     * Acquiring permissions of a specified weight is not supported in the
+     * spemaphore based implementation. Use {@link #acquirePermission()}
+     *
+     * @throws UnsupportedOperationException always for this implementation
+     */
+    @Override
+    public boolean acquirePermission(int weight) {
+        throw new UnsupportedOperationException("Acquiring permissions of a specified weight is not supported in the spemaphore based implementation");
+    }
+
+    /**
+     * Reserving permissions is not supported in the spemaphore based
+     * implementation. Semaphores are totally blocking by it's nature. So this
+     * non-blocking API isn't supported. Use {@link #acquirePermission()}
+     *
+     * @throws UnsupportedOperationException always for this implementation
      */
     @Override
     public long reservePermission() {
-        return -1;
```

Comment:
This is a non-backward compatible change. But I like the idea to throw a `UnsupportedOperationException`. What do you think @storozhukBM

Reply:
I can revert this but I doubt that any properly working code is depending on the current behaviour :)

Reply:
I could live with the change.

Reply:
I think this case is already covered in our documentation `amount of nanoseconds you should wait for reserved permission. if negative, it means you failed to reserve`
I'm not against such change in principle, but throwing exception can potentially eliminate some JIT optimizations, so I'd ask @walec51 to perform and post benchmarks and check if there are any regressions.

Reply:
After reading the documentation I think most people will assume that `failed to reserve` means that it failed because the calculated wait time was more then the timeout. Calling an unsupported function is a different calibre of failure to me. The first type of failure is expected to happen in normal production code, the second is not.

As for testing performance you mean that I should test if the performance of calling `AtomicRateLimiter#reservePermission` if affected by this change?

Because as for `SemaphoreBasedRateLimiter#reservePermission` I do not think performance is any issue because it should never be called in production code and the whole point of throwing the suggested exception is to prevent that as early in development as possible.

Reply:
> After reading the documentation I think most people will assume that failed to reserve means that it failed because the calculated wait time was more then the timeout. Calling an unsupported function is a different calibre of failure to me. The first type of failure is expected to happen in normal production code, the second is not.

I agree

Reply:
I tried running the JMH benchmarks but their Gradle configuration doesn't work on my machine.

```
cd resilience4j-ratelimiter
../gradlew --configure-on-demand -w -x check jmh
Configuration on demand is an incubating feature.
Processing 1 classes from /mnt/reszta/Projekty/WALCZAK IT/scr/resilience4j/resilience4j-ratelimiter/build/classes/java/jmh with "reflection" generator
Writing out Java source to /mnt/reszta/Projekty/WALCZAK IT/scr/resilience4j/resilience4j-ratelimiter/build/jmh-generated-sources and resources to /mnt/reszta/Projekty/WALCZAK IT/scr/resilience4j/resilience4j-ratelimiter/build/jmh-generated-resources
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.openjdk.jmh.util.Utils (file:/home/walec51/.gradle/caches/modules-2/files-2.1/org.openjdk.jmh/jmh-core/1.21/442447101f63074c61063858033fbfde8a076873/jmh-core-1.21.jar) to field java.io.PrintStream.charOut
WARNING: Please consider reporting this to the maintainers of org.openjdk.jmh.util.Utils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
# JMH version: 1.21
# VM version: JDK 11.0.4, OpenJDK 64-Bit Server VM, 11.0.4+11
# VM invoker: /usr/lib/jvm/java-11-openjdk-11.0.4.11-0.fc30.x86_64/bin/java
# VM options: <none>
# Warmup: 10 iterations, 10 s each
# Measurement: 10 iterations, 10 s each
# Timeout: 10 min per iteration
# Threads: 2 threads, will synchronize iterations
# Benchmark mode: Throughput, ops/time
# Benchmark: io.github.resilience4j.ratelimiter.RateLimiterBenchmark.atomicPermission

# Run progress: 0,00% complete, ETA 00:40:00
# Fork: 1 of 2
Error: Could not find or load main class org.openjdk.jmh.runner.ForkedMain
Caused by: java.lang.ClassNotFoundException: org.openjdk.jmh.runner.ForkedMain
<forked VM failed with exit code 1>
<stdout last='20 lines'>
</stdout>
<stderr last='20 lines'>
Error: Could not find or load main class org.openjdk.jmh.runner.ForkedMain
Caused by: java.lang.ClassNotFoundException: org.openjdk.jmh.runner.ForkedMain
</stderr>
```

any suggestions what could be the cause of this?

Reply:
I have a look at it. We did a major gradle update a while ago.